### PR TITLE
GitHub oauth + other goodies

### PR DESCRIPTION
**Pull request by lindelleric**
## In this pr: 

- Flow for authenticating with github. This is done via a OAuth app that lives in the avensia-oss organization. 
- Github helper function for finding the PR for the current branch. 
- Some extra helper functions in git.
- Better handling of env-file when none can be found in the file system. 
- Better handling of values for the prompts. They are now generic. 
- Pre-release version tag can now have numbers in them. This allows for a pre-release tag with the following format: `package@1.0.0-tag5.1`

### CODE REVIEW

File: src/git.ts [hunk 1]
```diff
@@ -70,19 +84,25 @@ export async function getDetailedTags(cwd?: string) {
...
-
-  const tags = await git(args, cwd);
-  const splitTags = tags.split('\n').map(s => s.trim());
-
-  let result = splitTags;
+  const tags = await tagListWithPattern(`${name}@*`, cwd);
+  let result = tags;
   if (excludeRC) {
-    result = splitTags.filter(s => !s.match(/(-rc.)/));
+    result = tags.filter(s => !s.match(/(-rc.)/));
```
Comment (RobertBorg) [inline src/git.ts]: i think u want to escape the period here.

File: src/github-access.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        await GithubAccess.captureCredentials();
+      } catch (e) {
+        throw new Error(`Failed to capture github credentials: ${e}`);
+      }
+    }
+  }
+
+  private static async captureCredentials() {
+    const promise = new Promise((resolve, reject) => {
+      const nonce = String(Math.random()).split('.')[1];
```
Comment (RobertBorg) [inline src/github-access.tsx]: you should use a cryptographically secure random source here.

File: src/github-access.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        if (code) {
+          if (state === nonce) {
+            res.statusCode = 200;
+            res.setHeader('content-type', 'text/html');
+            res.end('Validation OK! You can now close this tab. <script>setTimeout(() => close(), 1000)</script>');
+
+            server.close();
+            // Fetch credentials using temporary code.
+            const authUrl = `https://github.com/login/oauth/access_token?client_id=${GithubAccess.OAuthClientId}&client_secret=${GithubAccess.OAuthClientSecret}&code=${code}`;
```
Comment (RobertBorg) [inline src/github-access.tsx]: i don't agree it's safe to include client secret which is distributed to everyone. we should use this oauth flow https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce

**Review by RobertBorg — CHANGES_REQUESTED** [review]

File: src/git.ts (hunk 1 above)
Comment (lindelleric) [reply src/git.ts]: Yes, thanks :+1:

File: src/github-access.tsx (hunk 2 above)
Comment (lindelleric) [reply src/github-access.tsx]: Sure

File: src/github-access.tsx (hunk 3 above)
Comment (lindelleric) [reply src/github-access.tsx]: I agree that it looks weird to include the client secret. I followed github's [guide for authorizing oauth apps](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps), with inspiration from the auth flow in [github's cli](https://github.com/cli/cli/blob/6a8deb1f5a9f2aa0ace2eb154523f3b9f23a05ae/internal/authflow/flow.go#L25).

Comment (lindelleric) [reply src/github-access.tsx]: Changed the auth flow to use device-auth instead of client_secret!

**Comment by RobertBorg** [comment]
Works Great!

**Review by RobertBorg — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@avensia-oss/garn",
-  "version": "0.5.1",
+  "version": "0.5.2-rc.11",
```
Comment (csi-lund) [inline package.json]:
```suggestion
  "version": "0.5.2",
```

**Pull request merged**